rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,242 @@
+
+We can further improve the developer experience of rolling out new features gradually by automating the process.
+In the "Feature Flags" section of Cosmo Studio, we can add a function that will automatically roll out a new feature in steps from
+1%-10% in 1% increments, then 10%-30% in 5% increments, and finally 30%-100% in 10% increments.
+
+This function will increase the percentage of traffic that is sent to the new Subgraph based on the performance, correctness, and error rate of the new "feature" compared to the existing solution. If a threshold is exceeded, the function will stop the rollout and disable the feature flag.
+
+What this would allow teams to do is to set up a new feature flag,
+and then let the system automatically roll out the feature without requiring the platform team to deploy extra infrastructure or handle the rollout manually.
+
+The goal here is to avoid the platform team becoming a bottleneck for rolling out new features,
+but instead, to empower the teams to roll out new features safely and quickly,
+and to provide them but also the platform team with the necessary tools to monitor and observe rollouts.
+
+As you scale your API across more and more teams,
+we want to make sure that the platform team can focus on the platform itself,
+and not doing hand-holding for every new feature rollout,
+which should be a self-service operation.
+
+## Testing Experimental Features: The Solution
+
+Here's the default cmd to publish a new Subgraph Schema:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4001/graphql --label=team=A
+```
+
+Let's say, we'd like to introduce a new feature in the staging environment that will change the behavior of the API dramatically.
+We don't want to affect other developers who might be integration testing other features in the staging environment.
+As such, we want to publish a new Subgraph Schema under a feature flag.
+
+We can add a `--feature-flag` flag to the `subgraph publish` command:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4002/graphql --label=team=A --feature-flag=experimental --disable-composition-checks
+```
+
+We've deployed a copy of the API to a different port, and we've published a new Subgraph Schema under the feature flag `products.experimental`.
+We've also disabled composition checks for this Subgraph, as we're testing a new feature that might not be compatible with changes that other developers are making.
+
+To be able to test this new feature end-to-end, we need to enable the feature flag in our client.
+We can achieve this by setting a cookie:
+
+```bash
+curl -X POST http://localhost:4002/graphql -H "Content-Type: application/json" -H "Cookie: X-WG-Feature-Flag=products.experimental" --data '{ "query": "{ products { id name } }" }'
```

Comment:
This is a request to the router correct? Port `4002` was used by the subgraph above.